cobc/replace.c

```diff
@@ -86,6 +93,152 @@
 #define DEBUG_REPLACE
 #endif
 
+/* BEGIN implementation of a queue of text and token pairs. The
+   implementation could easily be translated for any other data */
+
+struct cb_token_queue {
+	int maxsize;
+        int pos;
+	int length;
+	const char **texts;
+	const char **tokens;
+};
+
+/* initial_size should be a power of two */
+static struct cb_token_queue *token_queue_new(int initial_size)
+{
+	struct cb_token_queue *q ;
+	q = cobc_malloc (sizeof (struct cb_token_queue));
+        q->maxsize = initial_size;
+        q->pos = 0;
+        q->length = 0;
+	q->texts = cobc_malloc (sizeof(char*) * initial_size * 2);
+        q->tokens = q->texts + initial_size;
+	return q;
+}
+
+static void token_queue_put(struct cb_token_queue *q, int strdup,
+                            const char *text,
+                            const char *token)
+{
+	int pos;
+        if (q->length == q->maxsize) {
+		int maxsize = q->maxsize * 2;
+		int n = q->maxsize - q->pos;
+                const char **p =
+			cobc_malloc (sizeof(char *) * maxsize * 2);
+                const char **old_text = q->texts;
+
+                memcpy (p, q->texts + q->pos,
+			sizeof(char *) * n);
+                if (q->pos > 0) {
+			memcpy(p + n, q->texts,
+			       sizeof(char *) * ( q->maxsize - n ) );
+                }
+                q->texts = p;
+
+                p += maxsize;
+
+                memcpy (p, q->tokens + q->pos,
+			sizeof(char *) * n);
+                if (q->pos > 0) {
+			memcpy(p + n, q->tokens,
+			       sizeof(char *) * ( q->maxsize - n ) );
+                }
+                q->tokens = p;
+
+		cobc_free (old_text);
```

Comment:
Wouldn't `realloc` be preferable here?

Reply:
No, because we need to put back the structure of the circular-buffer in a bigger space, and realloc will just copy the beginning.